{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select Top Features\n",
    "\n",
    "Technically, any forms of observation-by-feature matrix is acceptable for the method we developed, and users are\n",
    "encouraged to explore the usability of our method with other types of data, even not in a biological context.\n",
    "However, single-cell transcriptomics data, as provided, usually is of high dimensionality and contains technical and\n",
    "biological noise. With testing different approaches of reducing the dimensionality and noise, we recommend that users\n",
    "select a number of top differentially expressed genes (DEGs) for each cluster (or group of clusters) that a vertex represents.\n",
    "\n",
    "We implemented a fast Wilcoxon rank-sum test method which can be invoked with function [`select_top_features`](../generated/CytoSimplex.select_top_features).\n",
    "The test is done in a one group versus all other groups manner. Here, we will choose the top DEGs for Osteoblast cells \n",
    "(shortened as `\"OS\"`), Reticular cells (`\"RE\"`) and Chondrocytes (`\"CH\"`), as also shown in the previously mentioned \n",
    "publication. The number of top DEGs for each cluster is set to 30 (nTop = 30), thus 90 unique genes at maximum are expected to be returned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/wangych/Library/r-miniconda-arm64/lib/python3.10/site-packages/CytoSimplex/_select_top_features.py:228: RuntimeWarning: invalid value encountered in true_divide\n",
      "  z = z / usigma\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Steap1',\n",
       " 'Smim5',\n",
       " 'H2-DMa',\n",
       " 'Zcchc5',\n",
       " 'Lims2',\n",
       " 'Fam89a',\n",
       " 'Ninj2',\n",
       " 'Scin',\n",
       " 'Pygl',\n",
       " 'Slc2a5']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import CytoSimplex as csx\n",
    "import scanpy as sc\n",
    "adata = sc.read(filename='test.h5ad',\n",
    "                backup_url=\"https://figshare.com/ndownloader/files/41034857\")\n",
    "vertices = {\"OS\": \"Osteoblast_1\",\n",
    "            \"RE\": \"Reticular_1\",\n",
    "            \"CH\": \"Chondrocyte_1\"}\n",
    "selected_genes = csx.select_top_features(adata, cluster_var=\"cluster\", vertices=vertices, n_top=30)\n",
    "selected_genes[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, users can set `return_stats=True` to obtain a table of\n",
    "full Wilcoxon rank-sum test statistics including the result for all clusters, instead of selected vertices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/wangych/Library/r-miniconda-arm64/lib/python3.10/site-packages/CytoSimplex/_select_top_features.py:228: RuntimeWarning: invalid value encountered in true_divide\n",
      "  z = z / usigma\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>avgExpr</th>\n",
       "      <th>logFC</th>\n",
       "      <th>ustat</th>\n",
       "      <th>auc</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "      <th>pct_in</th>\n",
       "      <th>pct_out</th>\n",
       "      <th>feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CH</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3010.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CH</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3010.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Sox17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CH</td>\n",
       "      <td>10.610097</td>\n",
       "      <td>7.637897</td>\n",
       "      <td>4547.5</td>\n",
       "      <td>0.755273</td>\n",
       "      <td>4.666025e-08</td>\n",
       "      <td>4.881361e-07</td>\n",
       "      <td>81.481481</td>\n",
       "      <td>21.524664</td>\n",
       "      <td>Mrpl15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CH</td>\n",
       "      <td>6.077630</td>\n",
       "      <td>3.874176</td>\n",
       "      <td>3826.0</td>\n",
       "      <td>0.635443</td>\n",
       "      <td>9.406324e-04</td>\n",
       "      <td>3.637987e-03</td>\n",
       "      <td>48.148148</td>\n",
       "      <td>16.143498</td>\n",
       "      <td>Lypla1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CH</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.057590</td>\n",
       "      <td>2997.0</td>\n",
       "      <td>0.497758</td>\n",
       "      <td>7.669166e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.448430</td>\n",
       "      <td>Gm37988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242911</th>\n",
       "      <td>Reticular_2</td>\n",
       "      <td>1.582342</td>\n",
       "      <td>-0.496039</td>\n",
       "      <td>1048.0</td>\n",
       "      <td>0.483172</td>\n",
       "      <td>7.931468e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>11.111111</td>\n",
       "      <td>14.937759</td>\n",
       "      <td>PISD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242912</th>\n",
       "      <td>Reticular_2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.912214</td>\n",
       "      <td>846.0</td>\n",
       "      <td>0.390041</td>\n",
       "      <td>1.202520e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21.991701</td>\n",
       "      <td>DHRSX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242913</th>\n",
       "      <td>Reticular_2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.154048</td>\n",
       "      <td>1071.0</td>\n",
       "      <td>0.493776</td>\n",
       "      <td>7.746696e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.244813</td>\n",
       "      <td>CAAA01147332.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242914</th>\n",
       "      <td>Reticular_2</td>\n",
       "      <td>16.142743</td>\n",
       "      <td>2.744758</td>\n",
       "      <td>1350.0</td>\n",
       "      <td>0.622407</td>\n",
       "      <td>2.151089e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>83.402490</td>\n",
       "      <td>tdT-WPRE_trans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242915</th>\n",
       "      <td>Reticular_2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1084.5</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CreER-WPRE_trans</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>242916 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              group    avgExpr     logFC   ustat       auc          pval  \\\n",
       "0                CH   0.000000  0.000000  3010.5  0.500000           NaN   \n",
       "1                CH   0.000000  0.000000  3010.5  0.500000           NaN   \n",
       "2                CH  10.610097  7.637897  4547.5  0.755273  4.666025e-08   \n",
       "3                CH   6.077630  3.874176  3826.0  0.635443  9.406324e-04   \n",
       "4                CH   0.000000 -0.057590  2997.0  0.497758  7.669166e-01   \n",
       "...             ...        ...       ...     ...       ...           ...   \n",
       "242911  Reticular_2   1.582342 -0.496039  1048.0  0.483172  7.931468e-01   \n",
       "242912  Reticular_2   0.000000 -2.912214   846.0  0.390041  1.202520e-01   \n",
       "242913  Reticular_2   0.000000 -0.154048  1071.0  0.493776  7.746696e-01   \n",
       "242914  Reticular_2  16.142743  2.744758  1350.0  0.622407  2.151089e-01   \n",
       "242915  Reticular_2   0.000000  0.000000  1084.5  0.500000           NaN   \n",
       "\n",
       "                padj      pct_in    pct_out           feature  \n",
       "0                NaN    0.000000   0.000000               Rp1  \n",
       "1                NaN    0.000000   0.000000             Sox17  \n",
       "2       4.881361e-07   81.481481  21.524664            Mrpl15  \n",
       "3       3.637987e-03   48.148148  16.143498            Lypla1  \n",
       "4       1.000000e+00    0.000000   0.448430           Gm37988  \n",
       "...              ...         ...        ...               ...  \n",
       "242911  1.000000e+00   11.111111  14.937759              PISD  \n",
       "242912  1.000000e+00    0.000000  21.991701             DHRSX  \n",
       "242913  1.000000e+00    0.000000   1.244813    CAAA01147332.1  \n",
       "242914  1.000000e+00  100.000000  83.402490    tdT-WPRE_trans  \n",
       "242915           NaN    0.000000   0.000000  CreER-WPRE_trans  \n",
       "\n",
       "[242916 rows x 10 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats = csx.select_top_features(adata, cluster_var=\"cluster\", vertices=vertices, return_stats=True)\n",
    "stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The returned table can be considered as the concatenation of the tables of all tests, with the column `group` indicating \n",
    "which cluster the test is primarily based on. For example, the 3rd row is the result of the test for gene \"Mrpl15\" in group \"CH\", \n",
    "which represents the original cluster \"Chondrocyte_1\", against all other clusters. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
